emukit/quadrature/kernels/integration_measures.py

```diff
@@ -0,0 +1,147 @@
+from typing import Tuple, List
+
+
+class IntegrationMeasure:
+    """An integration measure"""
+
+    def __init__(self, name: str):
+        """
+        :param name: Name of the integration measure
+        """
+        self.name = name
+
+    def compute_density(self, x: np.ndarray) -> np.ndarray:
+        """
+        Computes the density at point x
+        :param x: points at which density is computes, shape (num_points, dim)
+        :return: the density at x, shape (num_points, )
+        """
+        raise NotImplementedError
+
+    def get_box(self) -> List[Tuple[float, float]]:
+        """
+        Meaningful box-bounds around the measure. Outside this box, the measure should be virtually zero.
+
+        :return: box in which the measure lies. List of D tuples, where D is the dimensionality and the tuples contain
+        the lower and upper bounds of the box i.e., [(lb_1, ub_1), (lb_2, ub_2), ..., (lb_D, ub_D)]
+        """
+        raise NotImplementedError
+
+
+class UniformMeasure(IntegrationMeasure):
+    """The Uniform measure"""
+
+    def __init__(self, bounds: List[Tuple[float, float]]):
+        """
+        :param bounds: List of D tuples, where D is the dimensionality of the domain and the tuples contain the lower
+        and upper bounds of the box defining the uniform measure i.e., [(lb_1, ub_1), (lb_2, ub_2), ..., (lb_D, ub_D)]
+        """
+        super().__init__('UniformMeasure')
+
+        # checks if lower bounds are smaller than upper bounds.
+        for bounds_d in bounds:
+            lb_d, ub_d = bounds_d
+            if lb_d >= ub_d:
+                raise ValueError("Upper bound of uniform measure must be larger than lower bound. Found a pair "
+                                 "containing (" + str(lb_d) + ", " + str(ub_d) + ").")
+
+        self.bounds = bounds
+        # uniform measure has constant density which is computed here.
+        self.density = self._compute_density()
+
+    def _compute_density(self) -> float:
+        differences = np.array([x[1] - x[0] for x in self.bounds])
+        volume = np.prod(differences)
+
+        if not volume > 0:
+            raise NumericalPrecisionError("Domain volume of uniform measure is not positive. Its value is "
+                                          + str(volume) + ". It might be numerical problems...")
+        return np.float(1. / volume)
+
+    def compute_density(self, x: np.ndarray) -> np.ndarray:
+        """
+        Computes the density at point x
+        :param x: points at which density is computes, shape (num_points, dim)
+        :return: the density at x, shape (num_points, )
+        """
+        # check if points are inside the box
+        bounds_lower = np.array([b[0] for b in self.bounds])
+        bounds_upper = np.array([b[1] for b in self.bounds])
+        inside_lower = 1 - (x < bounds_lower)
+        inside_upper = 1 - (x > bounds_upper)
+        inside_upper_lower = (inside_lower * inside_upper).sum(axis=1) == x.shape[1]
```

Comment:
These are cool tricks, but mixing comparisons and operations in one line is hard to read. Example: "How to properly understand first line? What (x < bounds_lower) returns, is it an array of True/False or something else? 1 minus this array is what, an array or a number?" And so on. Can you rewrite this in a bit more straightforward-to-read way?

Reply:
I added an explanation. It should be clear now.